Performance: In-place operations and caching infrastructure

mgyoo86 opened the pull request:
## Problem
Heavy allocations in `find_coil_currents!` and related functions caused:
- Excessive memory footprint from repeated temporary array allocations
- Poor performance due to GC pressure and unnecessary copies
- Type instability and boxing overhead in hot paths

## Target
Optimization focused primarily on **`GS_IMAS_pf_active__coil`** type, the most performance-critical coil configuration in FUSE workflows (e.g., `ActorPFdesign`).

## Solution

### 1. In-place Function Interface
Implemented zero-allocation variants for all flux computation functions:
- `ψ!`, `dψ_dR!`, `dψ_dZ!`, `d2ψ_dZ2!` write directly to pre-allocated buffers
- Internal dispatcher eliminates per-coil allocations

### 2. Dual Cache System
Caches reusable computed data with intelligent invalidation to handle configuration changes.

- **Current Cache:** Stores `current_per_turn` from IMAS time-series to avoid redundant data lookups

- **Element Cache:** Pre-computes IMAS coil geometry (outlines, turns_with_sign)

### 3. Type Stability Improvements
- Typed parameters (`Bp_fac::T`, `current_per_turn::T`) eliminate boxing allocations
- Reduced unnecessary allocations from type instability and boxing overhead

## Results
- **>10x** memory footprint reduction
- **~5x** performance improvement
- Bit-wise identical numerical results (validated)

## Performance Benchmarks

Benchmarked using `FUSE.ActorPFdesign` (representative workflow from FUSE tutorial example):

### Execution Time

| Metric | Before | After | Speedup |
|--------|--------|-------|---------|
| Mean | 8.078 s ± 704 ms | 1.510 s ± 186 ms | **5.35×** |
| Median | 7.844 s | 1.453 s | 5.40× |

### Memory Allocation

| Metric | Before | After | Reduction |
|--------|--------|-------|-----------|
| Memory | 2.60 GiB | 177.54 MiB | **15.0×** |
| Allocations | 69,040,727 | 2,287,032 | 30.2× |

<details>
<summary>Detailed Results (Click to expand) </summary>

**Benchmark command:**
```julia
using FUSE, BenchmarkTools

dd= IMAS.dd()
ini, act = FUSE.case_parameters(:KDEMO);

ini.equilibrium.R0 = 6.5;
ini.equilibrium.B0 = 7.8;
act.ActorCoreTransport.model = :FluxMatcher;

FUSE.init(dd, ini, act);

# warm-up
FUSE.ActorPFdesign(dd, act)

# benchmark
@benchmark FUSE.ActorPFdesign($dd, $act) samples=10 seconds=200
```

**Before (master):**
```julia
BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
 Range (min … max):  7.709 s …   10.045 s  ┊ GC (min … max): 3.30% … 17.08%
 Time  (median):     7.844 s               ┊ GC (median):    3.71%
 Time  (mean ± σ):   8.078 s ± 703.710 ms  ┊ GC (mean ± σ):  5.57% ±  4.22%
 Memory estimate: 2.60 GiB, allocs estimate: 69040727.
```

**After (perf/in-place):**
```julia
BenchmarkTools.Trial: 10 samples with 1 evaluation per sample.
 Range (min … max):  1.411 s …    2.031 s  ┊ GC (min … max): 0.00% … 26.38%
 Time  (median):     1.453 s               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.510 s ± 185.991 ms  ┊ GC (mean ± σ):  3.55% ±  8.34%
 Memory estimate: 177.54 MiB, allocs estimate: 2287032.
```

</details>

mgyoo86 committed:
feat: add default COCOS constant for performance

Add MXHcocos11 constant to avoid repeated COCOS construction
in hot paths, reducing allocation overhead in flux calculations.

mgyoo86 committed:
feat: add integration function for raw vector coordinates

Implement integrate() overload that accepts outline_r and outline_z
vectors directly, avoiding IMAS outline object allocation overhead.
This enables zero-allocation integration when outline data is cached.

mgyoo86 committed:
perf: add _gfunc overload with explicit outline parameters

Add optimized _gfunc for IMASelement that accepts pre-extracted
outline_r, outline_z, and turns_with_sign parameters. This enables
IMAS coil caching by avoiding repeated outline() calls and reducing
allocation overhead in tight loops.

mgyoo86 committed:
perf: implement derived data caching for IMAS coils

Add DerivedCoilData struct and caching infrastructure to avoid
repeated outline() calls and reduce allocation in Green's function
calculations:

- Add DerivedCoilData{T} for cached outline and turns data
- Extend GS_IMAS_pf_active__coil with optional _derived field
- Add derive_coil_data!/clear_derived_data! for cache management
- Modify Green function dispatch to use cached data when available

This provides significant performance improvement for repeated
calculations on IMAS coils by eliminating redundant data extraction
and object construction overhead.

mgyoo86 committed:
feat: implement in-place flux computation functions

Add non-allocating variants of flux calculation functions that write
results into pre-allocated arrays:

- Add ψ!, dψ_dR!, dψ_dZ!, d2ψ_dZ2! public API
- Implement _pfunc! for AbstractSingleCoil and IMAScoil vectors
- Implement _pfunc! for MultiCoil vectors
- Implement _pfunc! for Image vectors

These functions eliminate allocation overhead in tight loops and
enable efficient batch calculations. Particularly beneficial for
control matrix construction and iterative solver algorithms.

mgyoo86 committed:
perf: optimize control matrix construction with in-place operations

Replace allocating broadcast operations with in-place variants in
define_A function:

- Use ψ! instead of ψ.() for flux control points
- Use dψ_dR! and dψ_dZ! for saddle control points
- Use ψ! for iso control point calculations

This significantly reduces allocation overhead during control matrix
construction, especially beneficial for large numbers of control
points and coils. Measured allocation reduction of >95% in typical
use cases.

mgyoo86 committed:
refactor: reorganize test suite structure

Split monolithic runtests.jl into modular test files:
- Keep data definitions in runtests.jl
- Move existing tests to runtests_basic.jl
- Use include() pattern for better test organization

This improves maintainability and allows easier addition of
new test categories without cluttering the main test file.

mgyoo86 committed:
test: add comprehensive in-place function tests

Add test suite verifying in-place flux functions produce identical
results to original implementations:

- Test ψ!, dψ_dR!, dψ_dZ!, d2ψ_dZ2! accuracy
- Test flux_on_grid! correctness
- Verify MultiCoil support
- Add allocation tests confirming performance benefits
- Test consistency with summed values

All tests verify exact equality (==) between in-place and original
versions, ensuring zero functional regression from performance
optimizations.

mgyoo86 committed:
feat: implement current caching infrastructure for IMAS coils

Add time-based caching for current_per_turn values to avoid repeated
IMAS.get_time_array calls. Includes:

- CurrentCache{T} struct to store time and current_per_turn pairs
- Cache management functions: has_cached_current, get_cached_current,
  cache_current!, clear_current_cache!
- Updated GS_IMAS_pf_active__coil to include _cached_current field
- Two-level caching: _derived for static element data, _cached_current
  for dynamic coil current
- Exact time matching for cache hits to ensure correctness
- Type-safe conversions to coil's native numeric type

Also removes debug @infiltrate call and adds IMASdd dependency.

mgyoo86 committed:
test: add cache infrastructure validation tests

Comprehensive tests for DerivedCoilData and CurrentCache structs:

- Struct creation and field access verification
- Type conversion behavior (Int/Float32 to Float64)
- Vector operations for derived data collections
- Exact time matching semantics for cache hits
- Cache helper function signature validation

Tests validate cache infrastructure independently without requiring
full IMAS coil initialization. Integration tests with actual IMAS
coils should be added to files with IMAS data structures.

All 42 tests pass successfully.

mgyoo86 committed:
perf: implement zero-allocation element cache with geometry-based invalidation

Major refactoring of IMAS coil caching infrastructure for performance:

Cache Structure Improvements:
- Rename DerivedCoilData → ElementCache for clarity
- Rename _derived → _elements_cache, _cached_current → _current_cache
- Make ElementCache mutable to enable in-place updates
- Remove Union{..., Nothing} types, use empty vectors and NaN instead

Geometry-Based Cache Invalidation:
- Add source_geometry field to track Rectangle (r, z, width, height)
- Implement is_cache_valid with multiple dispatch:
  * Single element: is_cache_valid(elem_cache, element)
  * Vector: is_cache_valid(elements_cache, elements)
  * Current: is_cache_valid(current_cache, time)
- Automatic invalidation when geometry or turns_with_sign changes

Zero-Allocation Updates:
- Add update_cache! for in-place ElementCache updates
- Reuse existing vectors with resize!/copyto! pattern
- Reduce allocations from 3+ per element to ~0 for updates

Simplified API:
- Single entry point: ensure_valid_elements_cache!
- Validates and updates only invalid elements in-place
- Clean separation: validation → update → compute

Performance Impact:
- Cache hit: 0 allocations (unchanged)
- Single element update: ~0 allocations (was 3+)
- Geometry changes automatically invalidate relevant caches

All tests passing (51/51).

mgyoo86 committed:
refactor: extract IMAS coil cache management to separate file

Improved code organization by separating cache implementation from
main coil functionality. Cache validation and update logic now in
dedicated imas_coils_cache.jl module.

Changes:
- Created src/imas_coils_cache.jl with all cache management functions
- Reduced src/imas_coils.jl from 426 to 299 lines (-127 lines)
- Integrated cache tests into main test suite (runtests_coil_cache.jl)
- All 51 cache infrastructure tests passing

The separation maintains clear dependencies: struct definitions in
main file, implementation details in cache module via include().

mgyoo86 committed:
perf: use getfield instead of getproperty in cache validation

Micro-optimization to reduce allocations in hot path cache validation
and update functions. Direct field access via getfield avoids property
accessor overhead.

Changes:
- is_cache_valid: use getfield for geometry.rectangle field access
- is_cache_valid: compare fields individually instead of NamedTuple creation
- update_cache!: use getfield for all rectangle property access
- Added explicit type annotations for better type stability

This reduces allocations during cache invalidation checks which occur
on every Green function call in quad mode.

mgyoo86 committed:
refactor: simplify weighting calculation in define_A function

mgyoo86 committed:
perf: add type annotations to core math functions

- Add type parameters to D_ellipk, D2_ellipk for better type inference
- Add type annotations to Green's function and derivatives
- Add type parameters to RZq and Jacobian for performance
- Enables better compiler optimizations and inlining

mgyoo86 committed:
test: remove unnecessary ElementCache test cases

- Remove single ElementCache validation test
- Remove redundant hasmethod test for ElementCache
- Keep essential cache validation tests for vectors and CurrentCache

mgyoo86 committed:
refactor: use optimized flux interface in define_A

- Convert kwarg-based calls to positional arg versions
- Pass Bp_fac as positional parameter instead of kwarg
- Reduces allocation overhead from keyword argument handling
- Aligns with optimized flux function signatures

mgyoo86 committed:
perf: optimize IMAS coil Green's function computation

- Add type annotations to GS_IMAS_pf_active__coil constructor
- Refactor _dispatch_to_gfunc to use cached data efficiently
- Add optimized current_per_turn with caching
- Add specialized _pfunc! for IMAS coils
- Optimize _gfunc with better type stability and reduced bounds checks

mgyoo86 committed:
perf: optimize flux functions with type-stable non-kwarg versions

- Add type-annotated versions of ψ!, dψ_dR!, dψ_dZ!, d2ψ_dZ2!
- Implement faster non-kwarg versions taking Bp_fac as positional arg
- Add zero-current optimization to skip computation
- Add specialized _pfunc! for different coil types
- Improves type stability and reduces allocation overhead

mgyoo86 committed:
perf: optimize ellipke_nonneg with binary tree dispatch

Replace linear if-elseif chain with balanced binary tree structure
for polynomial table selection. Reduces worst-case comparisons from
10 to 2-3, achieving ~30% speedup with predictable performance.

mgyoo86 committed:
test: add comprehensive elliptic integral correctness tests

Add runtests_ellipke.jl with extensive test coverage for elliptic
integral functions (ellipke, ellipke_nonneg, ellipk, ellipe).

Test coverage includes:
- Bit-level correctness for positive/negative m across all polynomial tables
- Edge cases: m=0, m=1, m near 1, m>1 domain errors
- Negative m transformation verification (x = m/(m-1) mapping)
- Boundary consistency between polynomial table ranges
- 287 test cases validating all code paths

Ensures optimized elliptic implementations maintain exact bit-level
equivalence with reference implementations.

mgyoo86 committed:
refactor: remove unused `ellipke_ori` function to streamline elliptic integral computations

mgyoo86 committed:
refactor: optimize elliptic integral computation

- Rename ellipke to ellipke_ori (temporary, for testing)
- Add ellipke_nonneg: optimized K(m) and E(m) for non-negative m
  - Performs range selection once instead of twice
  - Type-stable with explicit else clause
- Implement new ellipke as primary interface
  - Uses ellipke_nonneg for m >= 0
  - Transforms negative m to positive domain

- Clean up docstrings
  - Remove specific performance percentages
  - Add clear usage examples for positive/negative m
  - Improve cross-references between functions

mgyoo86 committed:
Merge branch 'perf/ellipke' into perf/in-place

mgyoo86 committed:
refactor: add Green dispatch for hot path

mgyoo86 committed:
fix: add error handling for missing 'turns_with_sign' field in element cache creation

mgyoo86 committed:
refactor: restore sum(iterator) pattern for numerical accuracy

Revert to sum(iterator) approach with closures in _gfunc and
_dispatch_to_gfunc to leverage pairwise summation for improved
floating-point accuracy. Performance is comparable to manual loops.

- Add @inbounds to closures for bounds check elimination
- Rename closures to `_integrand` for clarity
- Use direct 2D array indexing (not views) for optimal performance
- Special case handling for single element in _dispatch_to_gfunc

mgyoo86 committed:
refactor: remove unused integration functions to simplify codebase

mgyoo86 committed:
refactor: replace _dispatch_to_gfunc with _compute_with_cache for improved performance

mgyoo86 committed:
Merge branch 'master' into perf/in-place

mgyoo86 committed:
refactor: remove unused IMASdd dependency from Project.toml

bclyons12 commented on src/imas_coils.jl:
```diff
@@ -1,13 +1,79 @@
+        elements_cache::Vector{ElementCache{T1}},
+        current_cache::CurrentCache{T1}) where {T1<:Real,T2<:Real,T3<:Real,T4<:Real}
+        new{T1,T2,T3,T4}(imas, tech, time0, green_model, elements_cache, current_cache)
+    end
```
The first two inner constructors here should be outer constructors, since they change the number of arguments. The last one is just the default one, so won't be needed at all if the other ones are outer constructors. That said, you can simplify it all with [Base.@kwdef](https://docs.julialang.org/en/v1/base/base/#Base.@kwdef).

mgyoo86 committed:
refactor: relax type constraints in Green's functions for better compatibility

Replace `where {T<:Real}` constraints with `::Real` annotations in point-to-point
Green's functions. This improves compatibility with ForwardDiff and Measurements.jl
while maintaining performance through Julia's type specialization.

mgyoo86 committed:
refactor: convert inner constructors to outer constructors for GS_IMAS_pf_active__coil

Move 4-arg and 5-arg convenience constructors outside the struct definition,
letting Julia generate the default 6-arg constructor. This follows idiomatic
Julia patterns while maintaining backward compatibility with existing code.

mgyoo86 committed:
docs: add docstring for GS_IMAS_pf_active__coil with cache sync warning

Document the struct fields and add a clear warning about cache
synchronization requirements when modifying current_per_turn.
Users must use setproperty! or set_current_per_turn! to avoid
desync between the wrapper and underlying IMAS data.

mgyoo86 commented:
@bclyons12 
Thanks for the thorough review! I've addressed each point in the recent commits:

1. **Relaxed Type restrictions**: Addressed in 720090a - relaxed type constraints in Green's functions. 
2. **Inner → outer constructors**: Addressed in 4415646 - converted to outer constructors as you commented. 
3. **Added a docstring for `GS_IMAS_pf_active__coil`**: 
- `set_current_per_turn!(::AbstractSingleCoil, ...)` does `coil.current_per_turn = value`, which calls `setproperty!` and so updates the cache. 
- Added docstring (80e7999) warning not to modify coil.imas.current.data directly.

bclyons12 commented on src/control.jl:
```diff
@@ -858,7 +858,8 @@ function define_A(coils::AbstractVector{<:AbstractCoil};
         # Build matrix relating coil Green's functions to boundary points
-        A[i, :] .= ψ.(coils, r, z; Bp_fac)
+        # Call in-place version to reduce allocations
+        ψ!((@views A[i, :]), coils, r, z, Bp_fac)
```
```suggestion
        ψ!(@view(A[i, :]), coils, r, z, Bp_fac)
```

bclyons12 commented on src/imas_coils.jl:
```diff
@@ -104,35 +191,98 @@ end
+    if is_cache_valid(cache, time0)
+        return cache.current_per_turn::T
+    else
+        value = T(IMAS.get_time_array(coil.imas.current, :data, time0))::T
```
I'm not sure you need to cast it with `T()`, then specify the type explicitly `::T`.

bclyons12 reviewed (commented):
I understand (and can reproduce) the performance benefits of all this, but I'm still a bit terrified of caches getting out of sync and leading to bugs that are _very_ difficult to figure out. Maybe we should let `GS_IMAS_pf_active__coil` have a `_use_cache::Bool` field that is `false` by default. Then users can intentionally turn this on when they know it's safe. Throughout FUSE, this will be true and we can just set `_use_cache=true` when we create our coils from the `dd`.

bclyons12 commented:
Better yet would be if we could get rid of the allocations themselves in computing `outline`, `current_per_turn`, etc., then the cache wouldn't really be necessary, but I understand that that has proved difficult.

mgyoo86 commented on src/control.jl:
```diff
@@ -874,8 +875,11 @@ function define_A(coils::AbstractVector{<:AbstractCoil};
+        dψ_dR!((@views A[ir, :]), coils, r, z, Bp_fac)
+
+        # Call in-place version to reduce allocations
+        dψ_dZ!((@views A[iz, :]), coils, r, z, Bp_fac)
```
@jmcclena  I’ll update this PR, checking the compatibility with the current master branch and applying Brendan’s suggestions, and let you know!

mgyoo86 committed:
Merge branch 'master' into perf/in-place

mgyoo86 committed:
Address Brendan's comments
